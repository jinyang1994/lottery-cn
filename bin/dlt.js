const { LotteryDlt } = require('../lib')

const formatSelected = str => {
  try {
    const [redStr, blueStr] = str.split('-')
    const red = redStr.split(',')
    const blue = blueStr.split(',')

    if (red.length !== 5 || blue.length !== 2) throw new Error()
    return { red, blue }
  } catch {
    return null
  }
}

require('yargs').command(
  'dlt <selected>',
  '中国体育彩票超级大乐透',
  yargs =>
    yargs
      .positional('selected', {
        describe: '所选号码 - 例如：01,02,03,04,05-01,02'
      })
      .option('start', {
        describe: '查询起始期号 - 例如：22011',
        type: 'number'
      })
      .option('end', {
        describe: '查询截止期号 - 例如：22012',
        type: 'number'
      })
      .option('point-id', {
        describe: '查询指定期号 - 例如：22012',
        type: 'number'
      })
      .option('lowest', {
        describe: '过滤每期中奖金额 - 例如：5',
        type: 'number'
      }),
  ({ selected, start, end, pointId, lowest }) => {
    const { red, blue } = formatSelected(selected)
    const dlt = new LotteryDlt(red, blue)

    if (pointId) {
      dlt.check(pointId, lowest && { lowest })
    } else {
      dlt.check(start || null, end || null, lowest && { lowest })
    }
  }
)
