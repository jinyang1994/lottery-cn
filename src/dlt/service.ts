import axios from 'axios'

interface IPrize {
  prizename: string
  singlebonus: number
}

interface IPointResult {
  issueno: string
  number: string
  refernumber: string
  prize: Array<IPrize>
}

interface IHistoryPointResult {
  total: number
  start: number
  num: number
  list: Array<IPointResult>
}

// numbers type
export type LotteryRedNumber = [string, string, string, string, string]
export type LotteryBlueNumber = [string, string]
export type LotteryPointId = number | null

export interface LotteryPoint {
  id: number
  red: LotteryRedNumber
  blue: LotteryBlueNumber
  prizes: {
    [requirement: string]: number
  }
}

const BASE_URL = 'https://api.jisuapi.com/caipiao/'
// const API_KEY = '025fec680c503ef9'
const API_KEY = '2c1ad428ccaab0f5'

const format = ({
  issueno: id,
  number: red,
  refernumber: blue,
  prize: prizes
}: IPointResult) => {
  const formatedPrizes = Object.fromEntries(
    prizes.map(({ prizename, singlebonus }) => [prizename, singlebonus])
  )

  return {
    id: parseInt(id, 10),
    red: red.split(' ') as LotteryRedNumber,
    blue: blue.split(' ') as LotteryBlueNumber,
    prizes: {
      '5+2': formatedPrizes['一等奖'] || 1_000_000,
      '5+1': formatedPrizes['二等奖'] || 100_000,
      '5+0': formatedPrizes['三等奖'] || 10_000,
      '4+2': formatedPrizes['四等奖'] || 3_000,
      '4+1': formatedPrizes['五等奖'] || 300,
      '3+2': formatedPrizes['六等奖'] || 200,
      '4+0': formatedPrizes['七等奖'] || 100,
      '3+1': formatedPrizes['八等奖'] || 15,
      '2+2': formatedPrizes['八等奖'] || 15,
      '3+0': formatedPrizes['九等奖'] || 5,
      '1+2': formatedPrizes['九等奖'] || 5,
      '2+1': formatedPrizes['九等奖'] || 5,
      '2+0': formatedPrizes['九等奖'] || 5
    }
  }
}

export const getPoint = async (
  id?: LotteryPointId
): Promise<LotteryPoint | null> => {
  const { data } = await axios.get<{ result: IPointResult }>('query', {
    baseURL: BASE_URL,
    params: {
      appkey: API_KEY,
      caipiaoid: 14,
      issueno: id || undefined
    }
  })

  return data.result ? format(data.result) : null
}

export const getHistoryPoint = async (
  start?: LotteryPointId,
  end?: LotteryPointId
): Promise<LotteryPoint[]> => {
  const { data } = await axios.get<{ result: IHistoryPointResult }>('history', {
    baseURL: BASE_URL,
    params: {
      appkey: API_KEY,
      caipiaoid: 14,
      issueno: end || undefined,
      start: 0,
      num: start ? 20 : 1
    }
  })
  const { list } = data.result
  let history: LotteryPoint[] = list
    .map(format)
    .filter(({ id }) => !start || id >= start)
  const hasStartPoint = !start || !!history.find(({ id }) => id <= start)
  const hasEndPoint = !end || !!history.find(({ id }) => id === end)

  // TODO: API can't get 'end point' when setting end in params
  if (!hasEndPoint) {
    const endPoint = await getPoint(end)

    if (endPoint) history = [endPoint, ...history]
  }
  if (!hasStartPoint) {
    const firstPoint = history.reduce(
      (res, item) => (res.id < item.id ? res : item),
      history[history.length - 1]
    )

    history = [...history, ...(await getHistoryPoint(start, firstPoint.id))]
  }

  return history
}

export default getPoint
