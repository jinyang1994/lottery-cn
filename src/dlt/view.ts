import { table, Alignment } from 'table'
import 'colors'
import type {
  LotteryRedNumber,
  LotteryBlueNumber,
  LotteryPoint
} from './service'

const format = (
  numbers: string[],
  selected?: string[],
  setColor?: (num: string) => string
) => {
  if (!setColor || !selected?.length) return numbers.join(' ')
  const reg = new RegExp(selected.join('|'), 'g')
  return numbers.join(' ').replace(reg, match => setColor(match))
}

const toCurrencyString = (amount: number): string =>
  amount.toLocaleString('zh-CN', { style: 'currency', currency: 'CNY' })

export const cli = (
  history: Array<Omit<LotteryPoint, 'prizes'> & { reward: number }>,
  selected: { red: LotteryRedNumber; blue: LotteryBlueNumber },
  viewConfig = { lowest: 0 }
): string => {
  const list = history.filter(
    item => !viewConfig.lowest || item.reward >= viewConfig.lowest
  )
  const tableConfig = {
    border: {
      topBody: '─',
      topJoin: '┬',
      topLeft: '┌',
      topRight: '┐',

      bottomBody: '─',
      bottomJoin: '┴',
      bottomLeft: '└',
      bottomRight: '┘',

      bodyLeft: '│',
      bodyRight: '│',
      bodyJoin: '│',

      joinBody: '─',
      joinLeft: '├',
      joinRight: '┤',
      joinJoin: '┼'
    },
    spanningCells: [
      {
        col: 0,
        row: list.length + 1,
        colSpan: 3,
        alignment: 'center' as Alignment
      }
    ]
  }
  const totalReward = list.reduce((acc, cur) => acc + cur.reward, 0)
  const data = [
    ['期号', '中奖号码', '奖金'],
    ...list.map(({ id, red, blue, reward }) => {
      const formated = {
        red: format(red, selected.red, num => num.red),
        blue: format(blue, selected.blue, num => num.blue)
      }

      return [
        id,
        `${formated.red} - ${formated.blue}`,
        `${toCurrencyString(reward)}`
      ]
    }),
    [
      totalReward
        ? `恭喜你！奖金总计：${toCurrencyString(totalReward).yellow}`
        : '很遗憾！没有中奖',
      '',
      ''
    ]
  ]

  return table(data, tableConfig)
}

export default cli
