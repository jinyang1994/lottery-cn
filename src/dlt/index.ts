import intersection from 'lodash/intersection'
import type {
  LotteryRedNumber,
  LotteryBlueNumber,
  LotteryPoint,
  LotteryPointId
} from './service'
import { getHistoryPoint, getPoint } from './service'
import * as view from './view'

interface Config {
  reportTo: string
  lowest: number
  feedback: (
    data: Array<Omit<LotteryPoint, 'prizes'> & { reward: number }>
  ) => void
}

export default class LotteryDlt {
  public readonly red: LotteryRedNumber
  public readonly blue: LotteryBlueNumber
  private _config: Config = {
    reportTo: '',
    lowest: 0,
    feedback: data => {
      const output = view.cli(
        data,
        {
          red: this.red,
          blue: this.blue
        },
        this.viewConfig
      )

      console.log(output)
    }
  }

  constructor(
    red: LotteryRedNumber,
    blue: LotteryBlueNumber,
    userConfig: Partial<Config> = {}
  ) {
    this.red = red
    this.blue = blue
    this._config = { ...this._config, ...userConfig }
  }

  get viewConfig(): { lowest: number } {
    return { lowest: this._config.lowest }
  }

  public config(userConfig: Partial<Config> = {}): void {
    this._config = { ...this._config, ...userConfig }
  }

  public report(email: string): LotteryDlt {
    if (this._config.reportTo) return this
    return new LotteryDlt(this.red, this.blue, { reportTo: email })
  }

  public async check(
    start?: LotteryPointId,
    end?: LotteryPointId
  ): Promise<void> {
    const { feedback } = this._config
    const data =
      start !== undefined && end !== undefined
        ? await LotteryDlt.checkPeriod(start, end)
        : await LotteryDlt.checkPoint(start)

    if (data.length)
      feedback(
        data.map(point => ({ ...point, reward: this.getRewardValue(point) }))
      )
  }

  private static async checkPoint(
    id?: LotteryPointId
  ): Promise<LotteryPoint[]> {
    const point = await getPoint(id)

    if (!point) return []
    return [point]
  }

  private static async checkPeriod(
    start: LotteryPointId,
    end: LotteryPointId
  ): Promise<LotteryPoint[]> {
    const history = await getHistoryPoint(start, end)

    if (!history.length) return []
    return history
  }

  private getRewardValue(point: LotteryPoint) {
    const checked = {
      red: intersection(point.red, this.red),
      blue: intersection(point.blue, this.blue)
    }
    const reward = point.prizes[`${checked.red.length}+${checked.blue.length}`]

    return reward ?? 0
  }
}
