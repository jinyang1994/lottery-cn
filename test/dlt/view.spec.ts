import { expect } from 'chai'
import * as view from 'dlt/view'

describe('Test dlt view', function () {
  it('simple data', function () {
    const str = view.cli(
      [
        {
          red: ['01', '02', '03', '04', '05'],
          blue: ['01', '02'],
          id: 22011,
          reward: 200
        }
      ],
      { red: ['01', '02', '03', '04', '05'], blue: ['01', '02'] }
    )

    expect(str).to.matchSnapshot()
  })

  it('multiple data', function () {
    const str = view.cli(
      [
        {
          red: ['01', '02', '03', '04', '05'],
          blue: ['01', '02'],
          id: 22011,
          reward: 200
        },
        {
          red: ['01', '02', '09', '10', '12'],
          blue: ['03', '06'],
          id: 22010,
          reward: 10
        }
      ],
      { red: ['01', '02', '03', '04', '05'], blue: ['01', '02'] }
    )

    expect(str).to.matchSnapshot()
  })

  it('use lowest config', function () {
    const str = view.cli(
      [
        {
          red: ['01', '02', '03', '04', '05'],
          blue: ['01', '02'],
          id: 22011,
          reward: 200
        },
        {
          red: ['01', '02', '09', '10', '12'],
          blue: ['03', '06'],
          id: 22010,
          reward: 10
        }
      ],
      { red: ['01', '02', '03', '04', '05'], blue: ['01', '02'] },
      { lowest: 20 }
    )

    expect(str).to.matchSnapshot()
  })

  it('did not win', function () {
    const str = view.cli(
      [
        {
          red: ['01', '02', '03', '04', '05'],
          blue: ['01', '02'],
          id: 22011,
          reward: 0
        },
        {
          red: ['01', '02', '09', '10', '12'],
          blue: ['03', '06'],
          id: 22010,
          reward: 0
        }
      ],
      { red: ['11', '12', '13', '14', '15'], blue: ['11', '12'] },
      { lowest: 20 }
    )

    expect(str).to.matchSnapshot()
  })
})
