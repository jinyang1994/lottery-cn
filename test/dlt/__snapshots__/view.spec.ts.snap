// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test dlt view did not win 1`] = `
"┌──────┬──────────┬──────┐
│ 期号 │ 中奖号码 │ 奖金 │
├──────┴──────────┴──────┤
│    很遗憾！没有中奖    │
└────────────────────────┘
"
`;

exports[`Test dlt view multiple data 1`] = `
"┌───────┬────────────────────────┬─────────┐
│ 期号  │ 中奖号码               │ 奖金    │
├───────┼────────────────────────┼─────────┤
│ 22011 │ [31m01[39m [31m02[39m [31m03[39m [31m04[39m [31m05[39m - [34m01[39m [34m02[39m │ ¥200.00 │
├───────┼────────────────────────┼─────────┤
│ 22010 │ [31m01[39m [31m02[39m 09 10 12 - 03 06 │ ¥10.00  │
├───────┴────────────────────────┴─────────┤
│        恭喜你！奖金总计：[33m¥210.00[39m         │
└──────────────────────────────────────────┘
"
`;

exports[`Test dlt view simple data 1`] = `
"┌───────┬────────────────────────┬─────────┐
│ 期号  │ 中奖号码               │ 奖金    │
├───────┼────────────────────────┼─────────┤
│ 22011 │ [31m01[39m [31m02[39m [31m03[39m [31m04[39m [31m05[39m - [34m01[39m [34m02[39m │ ¥200.00 │
├───────┴────────────────────────┴─────────┤
│        恭喜你！奖金总计：[33m¥200.00[39m         │
└──────────────────────────────────────────┘
"
`;

exports[`Test dlt view use lowest config 1`] = `
"┌───────┬────────────────────────┬─────────┐
│ 期号  │ 中奖号码               │ 奖金    │
├───────┼────────────────────────┼─────────┤
│ 22011 │ [31m01[39m [31m02[39m [31m03[39m [31m04[39m [31m05[39m - [34m01[39m [34m02[39m │ ¥200.00 │
├───────┴────────────────────────┴─────────┤
│        恭喜你！奖金总计：[33m¥200.00[39m         │
└──────────────────────────────────────────┘
"
`;
